src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesCodeFixProvider.cs

```diff
@@ -0,0 +1,702 @@
+                    c => FixAsync(context.Document, diagnostic, context.CancellationToken),
+                    equivalenceKey: GetEquivalenceKey(preference, isRemovableAssignment)),
+                diagnostic);
+
+            return Task.CompletedTask;
+        }
+
+        private static bool IsForEachIterationVariableDiagnostic(Diagnostic diagnostic, Document document)
+        {
+            // Do not offer a fix to replace unused foreach iteration variable with discard.
+            // User should probably replace it with a for loop based on the collection length.
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var root = diagnostic.Location.SourceTree.GetRoot();
+            return syntaxFacts.IsForEachStatement(root.FindNode(diagnostic.Location.SourceSpan));
+        }
+
+        private static string GetEquivalenceKey(UnusedValuePreference preference, bool isRemovableAssignment)
+            => preference.ToString() + isRemovableAssignment.ToString();
+
+        private static string GetEquivalenceKey(Diagnostic diagnostic)
+        {
+            var preference = AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.GetUnusedValuePreference(diagnostic);
+            var isRemovableAssignment = AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.GetIsRemovableAssignmentDiagnostic(diagnostic);
+            return GetEquivalenceKey(preference, isRemovableAssignment);
+        }
+
+        protected abstract SyntaxNode UpdateNameForFlaggedNode(SyntaxNode node, SyntaxToken newName);
+        protected abstract SyntaxToken GetForEachStatementIdentifier(TForEachStatementSyntax node);
+        protected abstract TBlockSyntax GenerateBlock(IEnumerable<TStatementSyntax> statements);
+        protected abstract ILocalSymbol GetSingleDeclaredLocal(TLocalDeclarationStatementSyntax localDeclaration, SemanticModel semanticModel, CancellationToken cancellationToken);
+        protected abstract void InsertAtStartOfSwitchCaseBlock(TSwitchCaseBlockSyntax switchCaseBlock, SyntaxEditor editor, TLocalDeclarationStatementSyntax declarationStatement);
+
+        private bool NeedsToMoveNewLocalDeclarationsNearReference(string diagnosticId) => diagnosticId == IDEDiagnosticIds.ValueAssignedIsUnusedDiagnosticId;
+        
+
+        protected override bool IncludeDiagnosticDuringFixAll(FixAllState fixAllState, Diagnostic diagnostic)
+        {
+            return fixAllState.CodeActionEquivalenceKey == GetEquivalenceKey(diagnostic) &&
+                !IsForEachIterationVariableDiagnostic(diagnostic, fixAllState.Document);
+        }
+
+        private IEnumerable<IGrouping<SyntaxNode, Diagnostic>> GetDiagnosticsGroupedByMember(
+            ImmutableArray<Diagnostic> diagnostics,
+            ISyntaxFactsService syntaxFacts,
+            SyntaxNode root,
+            out string diagnosticId,
+            out UnusedValuePreference preference,
+            out bool removeAssignments)
+        {
```

Comment:
We already have: ApplyMethodBodySemanticEditsAsync to do this sort of work.  IS there a reason we can't use that?

Reply:
Yes, primarily because there are post processing steps that need to apply changes to nodes which are different from the nodes on which diagnostics were reported. For example, `RemoveDiscardDeclarationsAsync` needs to remove discard declarations for the method bodies to which at least one diagnostic was reported. `MoveDeclarationNearReferenceAsync` needs to operate on new local declaration statements added by this fixer, not the original nodes on which diagnostic was reported. Additionally, the core code fix itself needs to fetch properties off the diagnostic while determining the fix, so the grouping is done at diagnostic level.

---
In reply to: [230604248](https://github.com/dotnet/roslyn/pull/30777#discussion_r230604248) [](ancestors = 230604248)